new download and releases page

fdoflorenzano opened the pull request.

SableRaf commented on i18n/react-intl/en.json:
```diff
@@ -1,6 +1,6 @@
 {
   "home": "Home",
-  "download": "Download",
+  "downloadTitle": "Create with code",
```
"Create with code, everywhere"

SableRaf commented on i18n/react-intl/en.json:
```diff
@@ -71,14 +71,24 @@
-  "earlierReleases": "Earlier releases have been removed because we can only support the current versions of the software. Read about <a href='https://github.com/processing/processing4/blob/main/build/shared/revisions.md'>revisions between individual releases</a>.",
-  "preReleases": "Pre-Releases",
-  "downloadOutro": "<a href='https://github.com/processing/processing4/wiki/Changes-in-4.0'>Read about changes in 4.0 since 3.x</a>.",
+  "downloadIntro": "With its vibrant community and endless resources, Processing is the perfect free and open-source tool for artists, designers, educators, and anyone looking to explore the creative side of coding. Download Processing and start creating today!",
```
"downloadIntro": "Processing is open source and is available for macOS, Windows, and Linux. Projects created with Processing are also cross-platform, and can be used on macOS, Windows, Android, Raspberry Pi, and many other Linux platforms."

SableRaf commented:
Looks great! Just a quick note: It looks like the tooltips are missing and the button design doesn't match the mockup (by the way, the tooltip should be on the Apple Silicon button instead of where it's placed in the mockup). I'll put the rest of the feedback in the code review.

Mockup:
![image](https://user-images.githubusercontent.com/290261/212943324-09e745ae-bd89-4165-b6ca-b781e5b98f33.png)

Current version:
![image](https://user-images.githubusercontent.com/290261/212942429-110ace80-c5be-4a6f-9914-9a995a760216.png)

SableRaf commented on src/hooks/download.js:
```diff
@@ -8,15 +8,28 @@ const getOS = (name) => {
-  else if (name.includes('macos-aarch64')) return '(Apple Silicon)';
-  else if (name.includes('linux-arm32')) return '(Raspberry Pi 32-bit)';
-  else if (name.includes('linux-arm64')) return '(Raspberry Pi 64-bit)';
+  if (name.includes('x64')) return 'Intel 64-bit';
```
Not sure if this is the place to change it but the order of the buttons for Linux should be as follow: Intel 64 at the top, then Rpi 32, then Rpi 64

SableRaf reviewed (changes requested):
Thanks Fernando! Here are some small changes needed. Also see the comment on the PR for some design specific feedback. The icon choices for the bottom links are good!

SableRaf commented:
Oh and one thing I forgot to mention: the latest release of [Processing is now 4.1.2](https://github.com/processing/processing4/releases/tag/processing-1290-4.1.2) (and no, it hasn't been updated on the website either 😅)

fdoflorenzano commented:
Hey @SableRaf !

I just pushes changes based on your comments:
- added the latest processing release 4.1.2
- tweaked the text downloadIntro and downloadTitle 
- added the sorting you asked for and applied to all assets, so no Intel is first in all of OS

I have some notes and questions for the couple pending comments:

> the button design doesn't match the mockup

It doesn't! This was a design update we did while I was implementing the mockup. We updated it cause the previous design for different download assets didn't visually suggest they are clickable or actionable. The OS images had a similar look and weren't clickable, so we went with a different and more obviously clickable look for them.

> It looks like the tooltips are missing 

Yes! I left it for later while developing and then forgot to ask about it. I'm not sure whats expected of the tooltip:
- Do all assets need a tooltip? Or just Apple Silicon?
- What's the text copy for each needed tooltip? Should it also have a translated version?
- What's the data in the bottom of the tooltip? Does that come from the GitHub release data?

SableRaf commented:
> It doesn't! This was a design update we did while I was implementing the mockup. We updated it cause the previous design for different download assets didn't visually suggest they are clickable or actionable. The OS images had a similar look and weren't clickable, so we went with a different and more obviously clickable look for them.

Thanks for the clarification! That brings up a tricky conceptual issue with the design. Ok, let's work through it :)

The OS images where meant to be clickable, similar to the [1Password download page](https://1password.com/downloads/). The thing is they can get away with it because they only have one possible download per OS, but in our case it complicates the UX flow. 

The version text below is similar to what you can find on the [Blender download page](https://www.blender.org/download/) where it is not clickable or actionable as you correctly pointed out. It's also meant to act as additional information for the download button itself so it is better suited as a sort of caption for it.

Alright, here's a proposed design that solves those issues by reorganizing the information and using radio buttons for selecting the versions:

![image](https://user-images.githubusercontent.com/290261/213266386-b5d7d2cb-84d2-4d0e-b633-1137c43b373f.png)

And with the tooltip:

![image](https://user-images.githubusercontent.com/290261/213266454-3ec8a709-fff4-4b16-a27a-b207e940dc5a.png)

The idea is to show the versions only for the currently selected OS. Note that I separated the Raspberry Pi releases for clarity and to emphasize the "runs on any platform" angle:

![image](https://user-images.githubusercontent.com/290261/213268553-aa59d858-1b99-4110-a630-cf054889f54a.png)

> Do all assets need a tooltip? Or just Apple Silicon?

Yes, all assets should have a tooltip.

> What's the text copy for each needed tooltip? Should it also have a translated version?

[Supported OS versions](https://github.com/processing/processing4/wiki/Supported-Platforms) + release date.

* Windows: "Tested on Windows 10. Untested on Windows 11 but should work."
* macOS Intel 64-bit: "Compatible with macOS 10.15.7 (Catalina), 11 (Big Sur), and 12 (Monterey)"
* macOS Apple Silicon: "Compatible with [M1 and M2 processors](https://support.apple.com/en-us/HT211814). Tested on macOS 12 (Monterey)"
* Linux 64-bit Intel: "Tested on Ubuntu 22.04."
* Raspberry Pi (32-bit ARM): "This is the only 32-bit platform that is supported with Processing 4."
* Raspberry Pi (64-bit ARM): "Not a ton of people are 64-bit with their RPi devices yet, but releases are happening."


> What's the data in the bottom of the tooltip? Does that come from the GitHub release data?

The timestamp for the release can be found in the JSON file for the release (e.g.: `processing-1289-4.1.1.json`) under `publishedAt`. The `md5` and `sha256` are not necessary right now.

fdoflorenzano commented:
Thanks for the text and comments! Will work on them during the week, and probably will have a talk with Rune and Dana about the designs.

Just to clarify about your proposal, what's the interaction flow you expect? What's clickable? What happens when clicked? Which clicks actually trigger a download?

SableRaf commented:
> Just to clarify about your proposal, what's the interaction flow you expect? What's clickable? What happens when clicked? Which clicks actually trigger a download?

Thanks for your question! There are two scenarios:

A) The OS detection picked the right version for the user: they click on the big blue button to trigger the download. Done.

B) The OS detection wasn't successful or the user needs the Apple Silicon, or RPi versions:
1. Click on the OS icon (the whole square area around the box should be clickable) for the desired system. Note: there should be a hover state for the OS button.
1. If needed and available, click on the radio button for the desired version (at the moment, this only applies to macOS and RPi)
1. Optional: click on the info button to see the tooltip
1. Click on the big blue button to trigger the download. Done

Note: The OS detection doesn't have to be perfect since the OS selector allows users to quickly switch to their desired system.

SableRaf commented:
Yeah @runemadsen I agree it's a bit of a tradeoff and the "surprise radio buttons" isn't the best. We could show the radio buttons at all time and make it also possible to click directly on the radio button to select the OS instead of the extra step of clicking the OS button then radio button, at the cost of more visual clutter.

Putting the download button below would improve the B flow but add friction to the A flow. Another tradeoff 🤔 Also it might make the placement of the "caption" a little more awkward. 

> I don't feel too strongly about the current layout, although I do think it's a pretty simple UX flow where you click the big blue button or click one of the more specific download links below. Everything is one click.

The concern I have over the current layout is that it brings back the "too many choices" issue the new design was meant to solve.

SableRaf commented:
Ok I tried to reorganize things. It brings a bunch of issues however:

1. When at the top, the download button acted as a headline of sorts (especially since the copy now says "Create with code.." instead of "Download Processing"). When bringing the button below, it gets more difficult to quickly grasp what this page is about.
1. The metadata under the button was colliding visually with the "older versions" text so I tried swapping it with the bottom buttons but the result doesn't flow as well as before.

![image](https://user-images.githubusercontent.com/290261/213495527-2bdeb9fe-658e-4900-805f-651ef82acf0d.png)

I would suggest keeping the vertical order as it was but keep the radio buttons visible even when the OS isn't selected, like so:

![image](https://user-images.githubusercontent.com/290261/213497548-dcbe1e61-ff3c-413c-b5f7-5a438e6771b8.png)

If you have other suggestions or ideas, let me know!

fdoflorenzano commented:
Thanks for the explanations! Going through the rationale behind the decisions made me think of another alternative.

<img width="1440" alt="Download" src="https://user-images.githubusercontent.com/13511560/213532290-b01592d3-8ee6-4d6d-9115-f7b2bdf20951.png">


<img width="1440" alt="Download-1" src="https://user-images.githubusercontent.com/13511560/213532296-d4f9d8ee-ad9f-4cf0-9184-7815e469545d.png">


This is trying to:
- Reduce the amount of choices proposed initially, so that flow A is well covered and users are pointed to it. To achieve this, other OS versions aren't shown initially, but are available after clicking the OS rectangles. The options are also buttons that directly start the download of the asset. 
- Reduce the amount of steps to also achieve flow B. Instead of using a single download button, each asset can be downloaded by directly clicking on it. So instead of 3 clicks, it would only bee two: select OS --> download asset. The big download button is also moved out of the OS section so that it's visually (and mentally) separated.

SableRaf commented:
That's a very good solution to avoid unexpectedly changing the state of the download button!

The buttons don't read clearly as download buttons though. How about something like this? And putting the info button outside of the download button to avoid nesting clickable areas?

---

![image](https://user-images.githubusercontent.com/290261/213538535-2ff2fd1b-fe32-4d69-9167-c86c79e752eb.png)

fdoflorenzano commented:
> ow about something like this? And putting the info button outside of the download button to avoid nesting clickable areas?

I like that! I'll go forward with this approach :)

SableRaf commented:
Thanks @fdoflorenzano :)

I explored the design a little further and I feel like there is a risk that many Apple Silicon users would download the Intel 64-bit version by mistake. I propose the following small tweak:

---

**Current design:** the Button says "Download Processing 4.1.2 for macOS". If you have an Apple Silicon Mac, you have to notice the small text that says "Intel 64-bit" and know what it means, or click the info button to maybe notice this is not the right version for you.

![image](https://user-images.githubusercontent.com/290261/213665950-1cd1f66d-7bad-4307-b89d-3fe5640dff02.png)

**Proposed tweak:** Add a line of text that says "Got an M1 or M2 CPU? Download the Apple Silicon version instead." with a link that downloads the Apple Silicon version.
![image](https://user-images.githubusercontent.com/290261/213667584-a46ca457-9a40-485e-bbb9-39fcc98e73fe.png)

fdoflorenzano commented:
@SableRaf I just pushed the latest tweaks including the warning for Apple Intel and mobile layouts!

SableRaf commented:
Thanks @fdoflorenzano! It looks amazing. Here are a few items of feedback:

**Main download button**
* The text should be slightly lower (`padding-top: 0.15em` on the `span` element seems to work). Same for the main download buttons
* The os-bit should also have a bit more room to breathe. `margin-top: 0.8em` on the `div` seems to do the trick.
* The button text should say "Download Processing 4.1.2 for Windows" or "Download Processing 4.1.2 for macOS", Linux, RPi (though we don't really need to detect Raspberry Pi)

Before
![image](https://user-images.githubusercontent.com/290261/214925726-85e462a7-37bb-4e63-814d-a571ef512b18.png)

After
![image](https://user-images.githubusercontent.com/290261/214925514-d4d8eccb-1318-417c-84c6-d3913a3af5bc.png)



**Tooltips**
* The color of the tooltip is the same as the download buttons, which creates a confusing visual experience. 
* The content could also use a little more padding.

![image](https://user-images.githubusercontent.com/290261/214916658-80b55e08-0a26-4eec-8487-f5c6720f8f6b.png)

**OS detection**
Would there be a way to also detect Apple Silicon? Even if it requires some feature detection that isn't 100% reliable, we could catch a number of users who may otherwise get the Intel version by mistake. cc @benfry

SableRaf commented:
I just got some updates for the tooltip text from Ben. I can make those changes myself later if you prefer. Just putting this here for reference.

**Windows:** 
"Compatible with Windows 10 and 11. For best compatibility, we recommend Windows 10."

**macOS Intel 64-bit:** 
"Compatible with macOS 10.15.7 (Catalina) through macOS 13 (Ventura). macOS 12 (Monterey) recommended."

**macOS Apple Silicon:** 
"Compatible with M1 and later processors. Where possible, we recommend macOS 12 (Monterey)."

**Linux 64-bit Intel:** 
"Tested and supported on Ubuntu 22.04, but will work with most major distributions."

**Raspberry Pi (32-bit ARM):** 
"Latest version of Raspberry Pi OS is recommended."

**Raspberry Pi (64-bit ARM):** 
"Latest version of Raspberry Pi OS is recommended. The 64-bit release Raspberry Pi OS is newer, so if you have issues, try the 32-bit version."

fdoflorenzano commented:
Thanks for the feedback @SableRaf! Just pushed another round of updates.



> **Main download button**

All considered and tweaked as proposed.



> * The color of the tooltip is the same as the download buttons, which creates a confusing visual experience.

Yeah that;s true, I added a light border so there's more of a visual separation to the background. 

> * The content could also use a little more padding.

I added a little bit more to the horizontal padding. 

I also tweaked a little z-index issue I found the tooltip buttons had with the donate banner; and fixed the positioning of the tooltip when too close to the right of the screen.

> **OS detection**
> Would there be a way to also detect Apple Silicon? Even if it requires some feature detection that isn't 100% reliable, we could catch a number of users who may otherwise get the Intel version by mistake.

I looked a bit into it and I couldn't find a well established and reliable way of doing it. Some approaches don't work in all browsers and even may break soon because the base API is subjected to change. There are some visual indicators for the Apple Silicon case to realize what they are downloading, and I'm also under the impression they usually have to check compatibility for their software.


> I just got some updates for the tooltip text from Ben.

Also updated these and their translation!

SableRaf reviewed (approved):
Tested locally and everything looks good to me! I recommend that we merge and include the new download page in the next release of the website. Thanks to Dana Elkis-Blanks, @fdoflorenzano, @runemadsen and the whole DSI team! ✨

fdoflorenzano commented:
@SableRaf amazing! Thanks so much for the feedback! it made for a better result at the end. 

Do you want me to merge or should I leave it to you to correctly time it? I'm not sure how you do it over here haha

SableRaf commented:
@fdoflorenzano I will merge it myself, thanks for asking! I'd like to ask for input on timing the release.
